// react-navigation-tabbar-collection/lib/commonjs/ColorfulTabBar/ColorfulTabBar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _TabColor = require("../assets/TabColor");

var _TabStyle = require("../assets/TabStyle");

var _Converter = require("../utils/Converter");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const ColorfulTabBar = _ref => {
  let {
    state,
    descriptors,
    navigation,
    maxWidth,
    height,
    darkMode = false,
    colorPalette = _TabColor.main_colors
  } = _ref;
  const BACKGROUND_COLOR = darkMode ? colorPalette.dark : colorPalette.light;
  const FOREGROUND_COLOR = darkMode ? colorPalette.light : colorPalette.dark;
  const ITEM_TOTAL = state.routes.length;
  return /*#__PURE__*/_react.default.createElement(_reactNative.SafeAreaView, {
    style: [_TabStyle.ColorfulStyle.container, {
      backgroundColor: BACKGROUND_COLOR,
      height: height
    }]
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [_TabStyle.ColorfulStyle.content, {
      maxWidth: maxWidth ? maxWidth : ITEM_TOTAL == 1 ? 150 : 100 * ITEM_TOTAL
    }]
  }, state.routes.map((route, index) => {
    const focusAnimation = (0, _react.useRef)(new _reactNative.Animated.Value(0)).current;
    const [width_text, setWidthText] = (0, _react.useState)(0);
    const [width_icon, setWidthIcon] = (0, _react.useState)(0);
    const {
      options
    } = descriptors[route.key];
    const label = options.tabBarLabel !== undefined ? options.tabBarLabel : options.label !== undefined ? options.label : options.title !== undefined ? options.title : route.name;
    const labelStyle = options.tabBarLabelStyle !== undefined ? options.tabBarLabelStyle : options.labelStyle !== undefined ? options.labelStyle : {};
    let color = options.tabBarActiveTintColor !== undefined ? options.tabBarActiveTintColor : options.color !== undefined ? options.color : colorPalette.primary;
    color = (color || '').charAt(0) == '#' ? color : colorPalette[color];
    const rgb_color = (0, _Converter.hexToRGB)(color);
    const icon = options.tabBarIcon !== undefined ? options.tabBarIcon : options.icon !== undefined ? options.icon : null;

    const renderIcon = focused => {
      if (icon === undefined || icon === null) {
        return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
          style: { ..._TabStyle.ColorfulStyle.itemIconNotFound,
            borderColor: focused ? color : FOREGROUND_COLOR
          }
        });
      }

      return icon({
        focused,
        color: focused ? darkMode ? FOREGROUND_COLOR : color : FOREGROUND_COLOR,
        size: 23
      });
    };

    const isFocused = state.index === index;

    const onPress = () => {
      const event = navigation.emit({
        type: 'tabPress',
        target: route.key,
        canPreventDefault: true
      });

      if (!isFocused && !event.defaultPrevented) {
        navigation.navigate(route.name);
      }
    };

    const onLongPress = () => {
      navigation.emit({
        type: 'tabLongPress',
        target: route.key
      });
    };

    (0, _react.useEffect)(() => {
      if (isFocused) {
        onFocusedAnimation();
      } else {
        notFocusedAnimation();
      }
    }, [isFocused]);

    const onFocusedAnimation = () => {
      _reactNative.Animated.timing(focusAnimation, {
        toValue: 1,
        duration: 500,
        useNativeDriver: false,
        easing: _reactNative.Easing.bezier(0.33, 1, 0.68, 1)
      }).start();
    };

    const notFocusedAnimation = () => {
      _reactNative.Animated.timing(focusAnimation, {
        toValue: 0,
        duration: 500,
        useNativeDriver: false,
        easing: _reactNative.Easing.bezier(0.33, 1, 0.68, 1)
      }).start();
    };

    const flexItem = focusAnimation.interpolate({
      inputRange: [0, 1],
      outputRange: [1, 2]
    });
    const filterColor = focusAnimation.interpolate({
      inputRange: [0, 1],
      outputRange: [`rgba(${rgb_color[0]}, ${rgb_color[1]}, ${rgb_color[2]}, 0)`, `rgba(${rgb_color[0]}, ${rgb_color[1]}, ${rgb_color[2]}, ${darkMode ? '1' : '.2'})`]
    });
    const HALF_ICON_WIDTH = width_icon / 2;
    const TEXT_ICON_WIDTH = width_text + HALF_ICON_WIDTH;
    const translateXIcon = focusAnimation.interpolate({
      inputRange: [0, 1],
      outputRange: [0, -TEXT_ICON_WIDTH / 2]
    });
    const translateXText = focusAnimation.interpolate({
      inputRange: [0, 1],
      outputRange: [0, HALF_ICON_WIDTH + 5]
    });
    const opacityText = focusAnimation.interpolate({
      inputRange: [0.3, 1],
      outputRange: [0, 1]
    });
    return /*#__PURE__*/_react.default.createElement(_reactNative.Animated.View, {
      key: index,
      style: [_TabStyle.ColorfulStyle.item, {
        flex: flexItem
      }]
    }, /*#__PURE__*/_react.default.createElement(_reactNative.TouchableOpacity, {
      accessibilityRole: "button",
      accessibilityState: isFocused ? {
        selected: true
      } : {},
      accessibilityLabel: options.tabBarAccessibilityLabel,
      testID: options.tabBarTestID,
      onPress: onPress,
      onLongPress: onLongPress,
      style: _TabStyle.ColorfulStyle.touchableItem
    }, /*#__PURE__*/_react.default.createElement(_reactNative.Animated.View, {
      style: [_TabStyle.ColorfulStyle.filterColor, {
        backgroundColor: filterColor
      }]
    }, /*#__PURE__*/_react.default.createElement(_reactNative.Animated.View, {
      style: {
        position: 'absolute',
        transform: [{
          translateX: translateXIcon
        }]
      },
      onLayout: e => {
        const {
          width
        } = e.nativeEvent.layout;
        setWidthIcon(width);
      }
    }, renderIcon(isFocused)), /*#__PURE__*/_react.default.createElement(_reactNative.Animated.View, {
      style: {
        position: 'absolute',
        opacity: opacityText,
        transform: [{
          translateX: translateXText
        }]
      },
      onLayout: e => {
        const {
          width
        } = e.nativeEvent.layout;
        setWidthText(width);
      }
    }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
      numberOfLines: 1,
      style: [_TabStyle.ColorfulStyle.itemText, {
        color: darkMode ? FOREGROUND_COLOR : color
      }, labelStyle]
    }, label)))));
  })));
};

var _default = ColorfulTabBar;
exports.default = _default;
